inputters/sil.lua

```diff
@@ -1,26 +1,50 @@
-local lpeg = require("lpeg")
+local base = require("inputters.base")
+
 local epnf = require("epnf")
 
-SILE.inputs.TeXlike = {}
+local sil = pl.class(base)
+sil._name = "sil"
+
+sil.order = 50
+
+sil.appropriate = function (round, filename, doc)
+  if round == 1 then
+    return filename:match(".sil$")
+  elseif round == 2 then
+    local sniff = doc:sub(1, 100)
+    local promising = sniff:match("\\begin") or sniff:match("\\document") or sniff:match("\\sile")
+    return promising and sil.appropriate(3, filename, doc)
+  elseif round == 3 then
+    local _parser = epnf.define(sil._grammar)
+    local status, tree = pcall(epnf.parsestring, _parser, doc)
+    local cmd = tree[1][1].command
+    return status and (cmd == "document" or cmd == "sile")
+  end
+end
+
+local bits = SILE.parserBits
 
-local ID = lpeg.C(SILE.parserBits.letter * (SILE.parserBits.letter + SILE.parserBits.digit)^0)
-SILE.inputs.TeXlike.identifier = (ID + lpeg.S":-")^1
 
-SILE.inputs.TeXlike.passthroughCommands = {
+sil.passthroughCommands = {
   ftl = true,
+  lua = true,
   math = true,
-  script = true
+  script = true,
+  sil = true,
+  xml = true
 }
-setmetatable(SILE.inputs.TeXlike.passthroughCommands, {
-    __call = function(self, command)
-      return self[command]
-    end
-  })
+
+function sil:_init (tree)
+  -- Save time when parsing strings by only setting up the grammar once per
+  -- instantiation then re-using it on every use.
+  self._parser = self:rebuildParser()
+  base._init(self, tree)
```

Comment:
Base init doesn't seem to have two args (i.e. one besides "self")
![image](https://user-images.githubusercontent.com/18075640/179837885-5faeda56-1830-4314-8f48-c2a6cd1d7d0c.png)

Reply:
It does have two args, `self` and `tree`. The call here is notably *not* using the Lua shortcut for passing self, it is calling the function from the parent instance (in this case `base`) but passing it the instantiation of the current class as `self`.

```lua
-- These two are equivalent in lua:
base:_init(tree)
base._init(base, tree)

-- This (which we are using) is different
base._init(self, tree)
```

The latter is different because at the time we pass it `base ~= self`.

Reply:
Hmm... Not sure why VSCode Lua checker hightlights this line then.... 
![image](https://user-images.githubusercontent.com/18075640/179966388-2378720f-a378-4395-b9e8-c753db56d2cc.png)

Ah, the base init is a noop, but it does expect one single argument indeed (= self).

Reply:
Correct, the base class is a NOOP and hence does not need the second argument. However VSCode might actually have a good point there, we should use `(_, _)` there just so even the unused prototype has the same number of arguments.

Reply:
Actually I'm full of it. That appears to be cruft and no longer need. The `_init()` function never uses the tree in any variant. It gets instantiated earlier then passed content to parse later.